homeassistant/components/osramlightify/light.py

```diff
@@ -4,270 +4,410 @@
+            self._xy_color = color_util.color_hs_to_xy(*self._hs_color)
+            self._luminary.set_rgb(*self._rgb_color, transition)
+        elif ATTR_RGB_COLOR in kwargs:
+            self._rgb_color = kwargs[ATTR_RGB_COLOR]
+            self._hs_color = color_util.color_RGB_to_hs(*self._rgb_color)
+            self._xy_color = color_util.color_RGB_to_xy(*self._rgb_color)
+            self._luminary.set_rgb(*self._rgb_color, transition)
+        elif ATTR_XY_COLOR in kwargs:
+            self._xy_color = kwargs[ATTR_XY_COLOR]
+            self._rgb_color = color_util.color_xy_to_RGB(*self._xy_color)
+            self._xy_color = color_util.color_xy_to_hs(*self._xy_color)
+            self._luminary.set_rgb(*self._rgb_color, transition)
+        elif ATTR_COLOR_NAME in kwargs:
+            self._rgb_color = color_util.color_name_to_rgb(
+                kwargs[ATTR_COLOR_NAME])
+            self._hs_color = color_util.color_RGB_to_hs(*self._rgb_color)
+            self._xy_color = color_util.color_RGB_to_xy(*self._rgb_color)
+            self._luminary.set_rgb(*self._rgb_color, transition)
 
         if ATTR_COLOR_TEMP in kwargs:
-            color_t = kwargs[ATTR_COLOR_TEMP]
-            kelvin = int(color_temperature_mired_to_kelvin(color_t))
-            self._luminary.set_temperature(kelvin, transition)
-
-        if ATTR_EFFECT in kwargs:
-            effect = kwargs.get(ATTR_EFFECT)
-            if effect == EFFECT_RANDOM:
-                self._luminary.set_rgb(
-                    random.randrange(0, 255), random.randrange(0, 255),
-                    random.randrange(0, 255), transition)
+            self._color_temp = kwargs[ATTR_COLOR_TEMP]
+            self._luminary.set_temperature(
+                int(color_util.color_temperature_mired_to_kelvin(
+                    self._color_temp)), transition)
+        elif ATTR_KELVIN in kwargs:
+            self._color_temp = color_util.color_temperature_kelvin_to_mired(
+                kwargs[ATTR_KELVIN])
+            self._luminary.set_temperature(kwargs[ATTR_KELVIN], transition)
+
+        self._is_on = True
+        if ATTR_BRIGHTNESS in kwargs:
+            self._brightness = kwargs[ATTR_BRIGHTNESS]
+            self._luminary.set_luminance(int(self._brightness / 2.55),
+                                         transition)
+        elif ATTR_BRIGHTNESS_PCT in kwargs:
+            self._brightness = int(kwargs[ATTR_BRIGHTNESS_PCT] * 2.55)
+            self._luminary.set_luminance(int(kwargs[ATTR_BRIGHTNESS_PCT]),
+                                         transition)
+        else:
+            self._luminary.set_onoff(True)
 
         self.schedule_update_ha_state()
 
     def turn_off(self, **kwargs):
         """Turn the device off."""
+        self._is_on = False
         if ATTR_TRANSITION in kwargs:
             transition = int(kwargs[ATTR_TRANSITION] * 10)
+            self._brightness = DEFAULT_BRIGHTNESS
```

Comment:
Why is this `DEFAULT_BRIGHTNESS` needed?

Reply:
If we turn off a light by setting its brightness to 0 (that transition actually does), it turns off the light and sets its brightness to 1%.  1% = 2 in HA format (`int(kwargs[ATTR_BRIGHTNESS_PCT] * 2.55)`).  Not to hardcode "2", I have moved it to a constant.

The reason to update properties here not waiting for `update()` is the same as for `turn_on()` method.